File: day_21/21.2.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"sync"
	"time"
)

type StartEnd struct {
	Start rune
	End   rune
}

type Keypad map[rune]map[rune]rune

var NUMERICAL_KEYPAD = Keypad{
	'A': {
		'^': '3',
		'<': '0',
	},
	'0': {
		'^': '2',
		'>': 'A',
	},
	'1': {
		'^': '4',
		'>': '2',
	},
	'2': {
		'^': '5',
		'v': '0',
		'<': '1',
		'>': '3',
	},
	'3': {
		'^': '6',
		'v': 'A',
		'<': '2',
	},
	'4': {
		'^': '7',
		'v': '1',
		'>': '5',
	},
	'5': {
		'^': '8',
		'v': '2',
		'<': '4',
		'>': '6',
	},
	'6': {
		'^': '9',
		'v': '3',
		'<': '5',
	},
	'7': {
		'v': '4',
		'>': '8',
	},
	'8': {
		'v': '5',
		'<': '7',
		'>': '9',
	},
	'9': {
		'v': '6',
		'<': '8',
	},
}

var DIRECTIONAL_KEYPAD = Keypad{
	'A': {
		'<': '^',
		'v': '>',
	},
	'^': {
		'>': 'A',
		'v': 'v',
	},
	'<': {
		'>': 'v',
	},
	'>': {
		'<': 'v',
		'^': 'A',
	},
	'v': {
		'<': '<',
		'>': '>',
		'^': '^',
	},
}

type Cache struct {
	data  map[string]string
	mutex sync.RWMutex
}

func NewCache() *Cache {
	return &Cache{
		data: make(map[string]string),
	}
}

func (c *Cache) Set(key string, value string) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	c.data[key] = value
}

func (c *Cache) Get(key string) (string, bool) {
	c.mutex.RLock()
	defer c.mutex.RUnlock()
	val, exists := c.data[key]
	return val, exists
}

func makeCodes(f *os.File) ([]string, error) {
	var codes []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		codes = append(codes, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return codes, nil
}

func findShortestPathLengthForCode(cache *Cache, numericalPaths map[StartEnd][]string, code string, maxDepth int) int {
	depth0 := numericalToDirectional(numericalPaths, code)
	var pathLengths []int
	for _, seq := range depth0 {
		pathLengths = append(pathLengths, pathLength(cache, seq, maxDepth))
	}
	return slices.Min(pathLengths)
}

func pathLength(cache *Cache, seq string, maxDepth int) int {
	if maxDepth == 0 {
		return len(seq)
	}

	stringDepth := int(math.Ceil(float64(maxDepth) / float64(2)))
	parts := splitByA(seq)
	for i := 0; i < stringDepth; i++ {
		for pIndex, s := range parts {
			parts[pIndex] = nextString(cache, s)
		}
	}

	remainingDepth := maxDepth - stringDepth
	total := 0
	for _, p := range parts {
		for _, s := range splitByA(p) {
			total += pathLength(cache, s, remainingDepth)
		}
	}
	return total
}

func nextString(cache *Cache, seq string) string {
	value, found := cache.Get(seq)
	if found {
		return value
	}

	result := ""
	for _, part := range splitByA(seq) {
		result += nextString(cache, part)
	}
	cache.Set(seq, result)
	return result
}

func findShortestPathLengthRecursive(directionalPaths map[StartEnd][]string, groupByAPaths map[string][]string, seq string, currentDepth int, maxDepth int) int {
	if currentDepth == maxDepth {
		return len(seq)
	}

	total := 0
	for _, s := range groupByA(seq) {
		nextSequences := directionalToDirectional(directionalPaths, groupByAPaths, s)
		var pathLengths []int
		for _, ns := range nextSequences {
			pathLengths = append(pathLengths, findShortestPathLengthRecursive(directionalPaths, groupByAPaths, ns, currentDepth+1, maxDepth))
		}
		total += slices.Min(pathLengths)
	}
	return total
}

func numericalToDirectional(numericalPaths map[StartEnd][]string, code string) []string {
	sequences := []string{""}
	for i, end := range code {
		var start rune
		if i == 0 {
			start = 'A'
		} else {
			start = rune(code[i-1])
		}

		paths := numericalPaths[StartEnd{start, end}]
		var newSequences []string
		for _, path := range paths {
			for _, s := range sequences {
				newSequences = append(newSequences, s+path)
			}
		}
		sequences = newSequences
	}
	return tidyUp(sequences) // TODO: maybe we shouldn't tidy
}

func directionalToDirectional(directionalPaths map[StartEnd][]string, groupByAPaths map[string][]string, directionalSeq string) []string {
	value, found := groupByAPaths[directionalSeq]
	if found {
		return value
	}

	sequences := []string{""}
	for i, end := range directionalSeq {
		var start rune
		if i == 0 {
			start = 'A'
		} else {
			start = rune(directionalSeq[i-1])
		}

		paths := directionalPaths[StartEnd{start, end}]
		var newSequences []string
		for _, path := range paths {
			for _, seq := range sequences {
				newSequences = append(newSequences, seq+path)
			}
		}
		sequences = newSequences
	}
	return tidyUp(sequences) // TODO: maybe we shouldn't tidy
}

func splitByA(seq string) []string {
	var aIndices []int
	for i, char := range seq {
		if char == 'A' {
			aIndices = append(aIndices, i)
		}
	}

	numAIndices := len(aIndices)
	if numAIndices < 2 {
		return []string{seq}
	}

	splitIndex := aIndices[numAIndices/2-1]
	firstHalf := seq[:splitIndex+1]
	secondHalf := seq[splitIndex+1:]
	return []string{firstHalf, secondHalf}
}

func groupByA(seq string) []string {
	var aIndices []int
	for i, char := range seq {
		if char == 'A' {
			aIndices = append(aIndices, i)
		}
	}

	var sequences []string
	start := 0
	for _, aIndex := range aIndices {
		sequences = append(sequences, seq[start:aIndex+1])
		start = aIndex + 1
	}
	return sequences
}

func _findShortestPaths(keypad Keypad, start rune, end rune, visited string) []string {
	for _, char := range visited {
		if char == start {
			return nil
		}
	}

	if start == end {
		return []string{"A"}
	}

	var paths []string
	for direction, nextStart := range keypad[start] {
		for _, path := range _findShortestPaths(keypad, nextStart, end, visited+string(start)) {
			paths = append(paths, string(direction)+path)
		}
	}

	if len(paths) == 0 {
		return paths
	}

	return tidyUp(paths)
}

func tidyUp(paths []string) []string {
	minLength := calculateMinPathLength(paths)
	var result []string
	for _, path := range paths {
		if len(path) == minLength {
			result = append(result, path)
		}
	}
	return result
}

func calculateComplexity(code string, length int) int {
	codeNum, _ := strconv.Atoi(code[:len(code)-1])
	return codeNum * length
}

func calculateMinPathLength(paths []string) int {
	return len(slices.MinFunc(paths, func(a, b string) int {
		return len(a) - len(b)
	}))
}

func preComputeKeypadPaths(keypad Keypad) map[StartEnd][]string {
	paths := make(map[StartEnd][]string)
	for start := range keypad {
		for end := range keypad {
			paths[StartEnd{start, end}] = _findShortestPaths(keypad, start, end, "")
		}
	}
	return paths
}

func preComputeGroupByAPaths(numericalPaths map[StartEnd][]string, directionalPaths map[StartEnd][]string) map[string][]string {
	groupByASequences := make(map[string]struct{})
	for _, sequences := range numericalPaths {
		for _, seq := range sequences {
			groupByASequences[seq] = struct{}{}
		}
	}
	for _, sequences := range directionalPaths {
		for _, seq := range sequences {
			groupByASequences[seq] = struct{}{}
		}
	}

	groupByAPaths := make(map[string][]string)
	emptyMap := make(map[string][]string)
	for seq := range groupByASequences {
		groupByAPaths[seq] = directionalToDirectional(directionalPaths, emptyMap, seq)
	}
	return groupByAPaths
}

func preComputeGroupByAWinners(directionalPaths map[StartEnd][]string, groupByAPaths map[string][]string) *Cache {
	maxDepth := 1 // TODO: change to see if we get a diff shortest path length
	winners := NewCache()
	for aSeq, depth0 := range groupByAPaths {
		var shortestPathLengths []int
		for _, seq := range depth0 {
			shortestPathLengths = append(shortestPathLengths, findShortestPathLengthRecursive(directionalPaths, groupByAPaths, seq, 0, maxDepth))
		}
		for i, seq := range depth0 {
			if shortestPathLengths[i] == slices.Min(shortestPathLengths) {
				winners.Set(aSeq, seq)
			}
		}
	}
	return winners
}

func main() {
	f, err := os.Open("C:/code/aoc-2024/day_21/day_21_input.txt")
	if err != nil {
		fmt.Println("Error opening file", err)
		return
	}

	defer f.Close()
	codes, err := makeCodes(f)
	if err != nil {
		fmt.Println("Error reading file:", err)
	}
	fmt.Println(codes)

	start := time.Now()

	numericalPaths := preComputeKeypadPaths(NUMERICAL_KEYPAD)
	directionalPaths := preComputeKeypadPaths(DIRECTIONAL_KEYPAD)
	groupByAPaths := preComputeGroupByAPaths(numericalPaths, directionalPaths)
	cache := preComputeGroupByAWinners(directionalPaths, groupByAPaths)

	depth := 2

	var complexities []int
	for _, code := range codes {
		complexities = append(complexities, calculateComplexity(code, findShortestPathLengthForCode(cache, numericalPaths, code, depth)))
	}

	total := 0
	for _, c := range complexities {
		total += c
	}
	fmt.Println(total)

	fmt.Println("took: ", time.Since(start))
}

File: day_21/21.2.py
import datetime


NUMERICAL_KEYPAD = {
    'A': {
        '^': '3',
        '<': '0',
    },
    '0': {
        '^': '2',
        '>': 'A',
    },
    '1': {
        '^': '4',
        '>': '2',
    },
    '2': {
        '^': '5',
        'v': '0',
        '<': '1',
        '>': '3',
    },
    '3': {
        '^': '6',
        'v': 'A',
        '<': '2',
    },
    '4': {
        '^': '7',
        'v': '1',
        '>': '5',
    },
    '5': {
        '^': '8',
        'v': '2',
        '<': '4',
        '>': '6',
    },
    '6': {
        '^': '9',
        'v': '3',
        '<': '5',
    },
    '7': {
        'v': '4',
        '>': '8',
    },
    '8': {
        'v': '5',
        '<': '7',
        '>': '9',
    },
    '9': {
        'v': '6',
        '<': '8',
    },
}

DIRECTIONAL_KEYPAD = {
    'A': {
        '<': '^',
        'v': '>'
    },
    '^': {
        '>': 'A',
        'v': 'v',
    },
    '<': {
        '>': 'v',
    },
    '>': {
        '<': 'v',
        '^': 'A'
    },
    'v': {
        '<': '<',
        '>': '>',
        '^': '^',
    },
}

def make_codes(contents):
    return [line for line in contents.split('\n') if line]

def find_shortest_paths(keypad, start_char, end_char, visited=None):
    if visited is None:
        visited = []

    if start_char in visited:
        return []

    if start_char == end_char:
        return [['A']]

    paths = []
    for direction, next_start_char in keypad[start_char].items():
        next_paths = [[direction] + path for path in find_shortest_paths(keypad, next_start_char, end_char, visited + [start_char])]
        paths += next_paths

    return [path for path in paths if len(path) == calculate_min_path_length(paths)]

def calculate_min_path_length(paths):
    if len(paths) == 0:
        return 0
    return min([len(path) for path in paths])

# update algorithm logic to keep track of min sequence length at each depth
# only fan out for sequences with the min sequence length
def find_all_sequences(code, depth):
    sequences = numerical_to_direction(code)

    for _ in range(depth):
        new_sequences = []
        for seq in sequences:
            new_paths = directional_transformation(seq)
            new_sequences.extend(new_paths)

        sequences = tidy_up(new_sequences)

    return sequences

def group_by_A(seq):
    a_indices = [i for i, char in enumerate(seq) if char == 'A']
    new_sequences = []
    start = 0
    for a_index in a_indices:
        new_sequences.append(seq[start : a_index + 1])
        start = a_index + 1
    return new_sequences

def numerical_to_direction(code):
    sequences = [[]]
    for i in range(len(code)):
        if i == 0:
            start = 'A'
        else:
            start = code[i - 1]
        end = code[i]
        paths = find_shortest_paths(NUMERICAL_KEYPAD, start, end)
        new_sequences = []
        for path in paths:
            for seq in sequences:
                new_sequences.append(seq + path)
        sequences = new_sequences
    return sequences

def directional_transformation(seq):
    # for group in group_by_A(seq):
    #     print(directional_to_directional(group))
    possible_sequences = directional_to_directional(seq)
    print(len(possible_sequences[0]))
    # return tidy_up(possible_sequences)

def combine(a, b):
    combos = []
    for a_elem in a:
        for b_elem in b:
            combos.append(a_elem + b_elem)
    return combos

def tidy_up(sequences):
    min_length = calculate_min_path_length(sequences)
    return [seq for seq in sequences if len(seq) == min_length]

def directional_to_directional(directional_seq):
    sequences = [[]]
    for i in range(len(directional_seq)):
        if i == 0:
            start = 'A'
        else:
            start = directional_seq[i - 1]
        end = directional_seq[i]
        paths = find_shortest_paths(DIRECTIONAL_KEYPAD, start, end)
        new_sequences = []
        for path in paths:
            for seq in sequences:
                new_sequences.append(seq + path)
        sequences = new_sequences
    return sequences

def calculate_complexity(code, sequences):
    return calculate_min_path_length(sequences) * int(code[:len(code) - 1])

if __name__ == '__main__':
    with open('21/day_21_test.txt', 'r') as f:
        contents = f.read()

    codes = make_codes(contents)

    # numerical_transform = ['<', 'A', '^', 'A', '>', '^', '^', 'A', 'v', 'v', 'v', 'A']
    # final = []
    # for group in group_by_A(['<', 'A', '^', 'A', '>', '^', '^', 'A', 'v', 'v', 'v', 'A']):
    #     first_dir_transform = directional_transformation(group)
    #     second_dir_transform = directional_transformation(first_dir_transform)
    #     final.extend(second_dir_transform)
    # print(len(final), final)

    # for group in group_by_A(['<', 'A', '^', 'A', '>', '^', '^', 'A', 'v', 'v', 'v', 'A']):
    #     print(f'group {group} -> {directional_to_directional(group)[0]}')

    # jeff = []
    # [jeff.extend(directional_to_directional(group)[0]) for group in group_by_A(['<', 'A', '^', 'A', '>', '^', '^', 'A', 'v', 'v', 'v', 'A'])]
    # print(f'{len(jeff)} - {jeff}')

    # jeff2 = []
    # [jeff2.extend(directional_to_directional(group)[0]) for group in group_by_A(jeff)]
    # print(f'{len(jeff2)} - {jeff2}')

    start = datetime.datetime.now()
    depth = 1
    print(sum([calculate_complexity(code, find_all_sequences(code, depth)) for code in codes]))
    print('took ', datetime.datetime.now() - start)
    # code = '029A'
    # for depth in range(10):
    #     print(f'{depth} - ', calculate_min_path_length(find_all_sequences(code, depth)))
    #     print('took ', datetime.datetime.now() - start)
    #     print()

# IDEAS:
#  - find all repeated subsequences within a sequence
#    - somehow group these together to minimize re-calculations
#    - build the next depth
#  - **CHECK THIS ONE OUT TOMORROW** rather than find repeated subsequences, use the 'A' to split the sequence up
#  - is this a math problem where we don't need to generate all combos, and instead can calculate size of smallest combo with a formula?
